### PR TITLE
Skips nulls and deleted in say

### PR DESCRIPTION
**Pull request by kriskog**
## About The Pull Request
Skips deleted and nulls in player_list instead of runtiming and breaking say for everyone. Not sure why we didnt do this ages ago, so if I am missing something obvious let me know?

## Why It's Good For The Game
Fix

## Changelog
:cl: Skoglol
fix: Say should be less likely to break when under heavy server strain.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/say.dm [hunk 1]
```diff
@@ -252,6 +252,8 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	var/list/the_dead = list()
 	for(var/_M in GLOB.player_list)
 		var/mob/M = _M
+		if(!M || QDELETED(M))
```
Comment (4dplanner) [inline code/modules/mob/living/say.dm]: #define QDELETED(X) (!X || QDELING(X))

**Review by Jalleo — APPROVED** [review]

File: code/modules/mob/living/say.dm (hunk 1 above)
Comment (kriskog) [inline code/modules/mob/living/say.dm]:
```suggestion
		if(QDELETED(M))

```

File: code/modules/mob/living/say.dm [hunk 2]
```diff
@@ -252,6 +252,9 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	var/list/the_dead = list()
 	for(var/_M in GLOB.player_list)
 		var/mob/M = _M
+		if(QDELETED(M))
+
```
Comment (kriskog) [inline code/modules/mob/living/say.dm]:
```suggestion
```

File: code/modules/mob/living/say.dm (hunk 1 above)
Comment (kriskog) [reply code/modules/mob/living/say.dm]: Thanks

**Comment by spookydonut** [comment]
It's problem hiding?

**Comment by kriskog** [comment]
We already got regular cleanup of nulls from these lists, which alleviates most of the issues. I still see it happen fairly often with already deleted but not yet removed mobs if there is a lot of lag and gc is running behind.

Performance concerns?

**Comment by optimumtact** [comment]
we didn't do this because it's not really a fix, and it just obscures a bug with byond's login and logout handling routines.

**Comment by ShizCalev** [comment]
@optimumtact would you prefer this merged or just closed then if it's inherently an ancient byond issue?

**Comment by optimumtact** [comment]
I have no real strong feelings on this one

File: code/modules/mob/living/say.dm [hunk 3]
```diff
@@ -252,6 +252,8 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	var/list/the_dead = list()
 	for(var/_M in GLOB.player_list)
 		var/mob/M = _M
+		if(QDELETED(M))
```
Comment (ShizCalev) [inline code/modules/mob/living/say.dm]: put a note saying this is a workaround for byond login / logout bugs, and to remove once fixed. likewise, put `see TG PR #49004` as well.

(aka likely never, but still)

**Comment by kriskog** [comment]
Allright comment added, seem okay?
Moved a source client check out of the loop while I was at it, should save us quite a few iterations.

**Pull request merged**